### PR TITLE
Port `glob_to_regex` and `re_word_boundary` from Synapse

### PR DESCRIPTION
**Pull request by squahtx**
Synapse:
https://github.com/matrix-org/synapse/blob/a82b90ab32629108fc69439a8cd38d025a406019/synapse/util/__init__.py#L190
https://github.com/matrix-org/synapse/blob/a82b90ab32629108fc69439a8cd38d025a406019/tests/util/test_glob_to_regex.py

Sygnal:
https://github.com/matrix-org/sygnal/pull/268/files

### CODE REVIEW

File: matrix_common/regex.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    Adds word boundary characters to the start and end of a pattern to require that the
+    match occur as a whole word.
+
+    If the start or end of the pattern is a non-word character, then a word boundary is
+    not required to precede or succeed it.
+
+    A word boundary is considered to be the boundary between a word and non-word
+    character. As such, the returned pattern is not appropriate for internationalized
+    text search because there are languages which do not use spaces between words.
```
Comment (squahtx) [inline matrix_common/regex.py:83]: Unfortunately it looks like we use this for matching words in message content?:
https://github.com/matrix-org/synapse/blob/develop/synapse/push/push_rule_evaluator.py#L187
https://github.com/matrix-org/synapse/blob/develop/synapse/push/push_rule_evaluator.py#L167

File: matrix_common/regex.py [hunk 2]
```diff
@@ -51,21 +51,35 @@ def glob_to_regex(glob: str, word_boundary: bool = False) -> Pattern[str]:
...
+
+    Returns:
+        A new pattern that only matches on whole words. The new pattern may match up to
+        one extra non-word character on either side. The exact match is provided by a
+        capture group.
     """
-    # we can't use \b as it chokes on unicode. however \W seems to be okay
-    # as shorthand for [^0-9A-Za-z_].
-    return r"(^|\W)%s(\W|$)" % (r,)
+    # `^|\W` and `\W|$` handle the case where `pattern` starts or ends with a non-word
```
Comment (DMRobertson) [inline matrix_common/regex.py:93]: Are these necessary? Python docs say

> Note that formally, \b is defined as the boundary between a \w and a \W character (or vice versa), or between \w and the beginning/end of the string.

File: matrix_common/regex.py [hunk 3]
```diff
@@ -42,21 +43,23 @@ def glob_to_regex(glob: str, word_boundary: bool = False) -> Pattern[str]:
...
 
     if word_boundary:
-        res = to_word_pattern(res)
+        pattern = to_word_pattern(pattern)
     else:
-        # \A anchors at start of string, \Z at end of string
-        res = r"\A" + res + r"\Z"
+        # `\A` anchors at start of string, `\Z` at end of string
+        # `\Z` is not the same as `$`! The latter will match the position before
+        # a `\n` at the end of the string.
```
Comment (DMRobertson) [inline matrix_common/regex.py:66]: Thanks for this. Is `\A` the same as `^`?

File: tests/test_regex.py [hunk 4]
```diff
@@ -44,16 +44,16 @@ def test_multi_wildcard(self) -> None:
         pat = glob_to_regex("**baz")
         self.assertTrue(pat.match("agsgsbaz"), "** should match any string")
         self.assertTrue(pat.match("baz"), "** should match the empty string")
-        self.assertEqual(pat.pattern, r"\A.{0,}baz\Z")
+        self.assertEqual(pat.pattern, r"\A(.{0,}baz)\Z")
```
Comment (DMRobertson) [inline tests/test_regex.py]: Corresponds to a change in 80cbe997be0cdf1e1dddbc7300de53f29039dbc8, I think?

File: tests/test_regex.py [hunk 5]
```diff
@@ -68,3 +68,22 @@ def test_ignore_case(self) -> None:
...
+
+class WordPatternTestCase(TestCase):
+    def test_whole_word(self) -> None:
+        """Tests matching on whole words."""
+        pattern = to_word_pattern("foo bar")
+
+        self.assertRegex("foo bar", pattern)
+        self.assertRegex(" foo bar ", pattern)
+        self.assertRegex("baz foo bar baz", pattern)
+        self.assertNotRegex("foo baré", pattern)
```
Comment (DMRobertson) [inline tests/test_regex.py]: Guessing the order of words in the test string matters here? If so, could add `self.assertNotRegex("bar foo", pattern")`

File: matrix_common/regex.py (hunk 2 above)
Comment (squahtx) [reply matrix_common/regex.py:93]: The thinking here was to have "abc%" match "hello abc%!".
ie. if the search string ends with a non-word character, it should match everywhere and not just on a `\b`.

The `^` and `$` are indeed unnecessary.

File: matrix_common/regex.py (hunk 3 above)
Comment (squahtx) [reply matrix_common/regex.py:66]: I'm not actually sure.

Comment (squahtx) [reply matrix_common/regex.py:66]: With `re.MULTILINE`, they're different:
```
>>> re.search("\Aabc$", "\nabc\n", re.MULTILINE)
>>> re.search("^abc$", "\nabc\n", re.MULTILINE)
<re.Match object; span=(1, 4), match='abc'>
```
without re.MULTILINE, I can't seem to find a difference.

File: tests/test_regex.py (hunk 4 above)
Comment (squahtx) [reply tests/test_regex.py]: Yep, that commit added a capture group.

File: matrix_common/regex.py (hunk 2 above)
Comment (squahtx) [reply matrix_common/regex.py:93]: > The `^` and `$` are indeed unnecessary.

Turns out removing them fails the tests. `^`/`$` is only considered a `\b` if the first or last character was a letter.
```py
>>> re.search("abc \b", "abc ")
>>> re.search("abc ($|\b)", "abc ")
<re.Match object; span=(0, 4), match='abc '>
```

File: tests/test_regex.py (hunk 5 above)
Comment (squahtx) [reply tests/test_regex.py]: It does, I've added another test.

**Comment by squahtx** [comment]
I added some descriptions for a few asserts so that what they're testing is clearer.

File: matrix_common/regex.py (hunk 2 above)
Comment (DMRobertson) [reply matrix_common/regex.py:93]: I think that makes sense. The end-of-line-input `$` isn't adjacent to the end of a word. I think we have a test_ends_with_non_letter covers this too?

**Review by DMRobertson — APPROVED** [review]
LGTM. Thanks!

File: matrix_common/regex.py (hunk 2 above)
Comment (squahtx) [reply matrix_common/regex.py:93]: Yep that's the test that fails.

**Pull request merged**